README
=pod

To better view this file, use:

	perldoc README

=head1 graphcnv

Take textual graph descriptions and turn them into something that looks pretty.

This package is intended to integrate Graph::Easy into Wikimedia, the software
behind popular wikis like http://en.wikipedia.org/.

For instance this input:

	[ Bonn ] -> [ Berlin ]
	[ Berlin -> [ Frankfurt ] { border: 1px dotted black; }
	[ Frankfurt ] -> [ Dresden ]
	[ Berlin ] -> [ Potsdam ]
	[ Potsdam ] => [ Cottbus ]

would be rendered in ASCII as:

	+------+     +--------+      .............     +---------+
	| Bonn | --> | Berlin |  --> : Frankfurt : --> | Dresden |
	+------+     +--------+      .............     +---------+
	               |
	               |
	               v
	             +---------+     +---------+
	             | Potsdam | ==> | Cottbus |
	             +---------+     +---------+

The HTML, SVG or PNG output would look similiar, except be more pretty :o)

=head1 Check the Signature

First you need to check the signature on the package to verify that
it is untampered and intact. To do this you have to:

=over 2

=item *

Install the Perl module Module::Signature from http://search.cpan.org

=item *

Get my key from http://bloodgate.com/tels.asc, import it into
GnuPG with:

	gpg --import tels.asc

You may also let the C<cpansign> utility fetch it automatically
from a keyserver, if that works for you.

=item *

Extract the package to a directory, change there and then:

=item *

Type on a console:

	cpansign --verify

inside the unpacked directory. 

=back

If the last step says "good signature" or "SIGNATURE VERIFIED OK",
everything is all right.

Note that if C<cpansign> let GnuPG download my key from a keyserver, it
might put a C<tels.asc> file in the package directory, and then afterward
complain that "tels.asc" is not in the MANIFEST:

 Not in MANIFEST: tels.asc
 ==> MISMATCHED content between MANIFEST and distribution files! <==

You can safely ignore this warning.

B<NOTE:> Please notify me immidiately if the signature
does not verify. In that case do B<NOT> install this software, your system
might get compromised!

=head1 Installation

Install the following Perl modules into your system:

	Heap
	Config-Tiny
	Graph-Easy
	Graph-Easy-As_svg	(optional, for SVG output)

You can get these modules from C<http://search.cpan.org/>.

Afterwards, you should be able to run this on any command line:

	perl -MGraph::Easy\ 99

and it should report your installed Graph::Easy version (like v0.34).

	Graph::Easy version 99 required--this is only version 0.34.
	BEGIN failed--compilation aborted.

If it says something about "Graph::Easy" not being found, you need to
install it first as root (see above!).

Note: You do NOT need to install Graph::Simple (the former name of
Graph::Easy), nor Graph (formerly used by Graph::Easy). If they are already
installed, that will not do any harm, however.

=head2 Run the testsuite

Run in the directory of this package:

	perl Makefile.PL
	make test

All tests should pass. If not, please notify me.

Then place the contents of this mediawiki-graph.tar.gz file so that it
is a subdirectory named 'graph' of your wiki installation like this:

	http_docroot
		\ - cgi-bin			etc
		\ - wiki
			\ - graph		<-- here
			\ - math		etc

Make a backup of your wiki directory! I repeat, MAKE A BACKUP! You have been
warned!

Now copy B<all> the files from the C<extensions/> subdirectory of C<graph> to
to the extension dir of your wiki, like so:

	http_docroot
		\ - cgi-bin			etc
		\ - wiki
			\ - graph
				\ - extensions	<-- from here
			\ - extensions		<-- to here

At present this is only one file, C<Graph.php>.

Then include the following line in your C<LocalSettings.php>:

	include('extensions/Graph.php');

Now you should be able to edit any page of the wiki and insert:

	<graph>
	[ Bonn ] -> [ Berlin ]
	</graph>

and hit preview.

If you get error messages like:

	Error running graphcnv.

make sure that you can run manually (from your wiki directory):

	./graph/graphcnv '[ Bonn ] -> [ Berlin ]' 'ascii'

and get a pretty ASCII art back.

=head1 CONFIGURATION

Inside the C<graph> subdirectory there is a file called C<graph.cfg.sample>.
Copy it to C<graph.cfg> like so:

	http_docroot
		\ - cgi-bin			etc
		\ - wiki
			\ - graph
				graph.cfg.sample	<-- from here
				graph.cfg		<-- to here

Then edit it to your match your settings.

For more information please see L<http://bloodgate.com/perl/graph/>
or send me an email.

=head1 SVG AND PNG OUTPUT

To enable SVG/PNG etc. output, you need to create a directory for the
files that will be generated for the graphs. (Inlining the SVG code in
the HTML page directly does not work, since no browser seesm to support
this properly at this time).

The C<.svg> (or C<.png>) files are written to the output directory
set in the config file, which defaults to C<./images/graph/>.
So you need to make sure that this directory exists and is writable by the
web server process:

	http_docroot
		\ - cgi-bin			etc
		\ - wiki
			\ - images
				\ - graph	<-- create this directory
			\ - extensions		etc

You need also make sure that you web server process can write to this
directory, so you might need to give it the same user/group as the
process is using, f.i. C<chmod nobody.nogroup images/graph>.

=head1 AUTHOR

Copyright (C) 2004 - 2006 by Tels http://bloodgate.com/

Contact: <nospam-abuse@bloodgate.com> (literally :)

This library is free software; you can redistribute it and/or modify
it under the same terms of the GPL version 2.



CHANGES

2006-01-28 v0.18 Tels 20 tests
  * require Graph::Easy 0.40
  * fix an unquoted bareword in Graph.php (Thanx Mutante!)
  * fix Makefile.PL to work with newer YAML

2005-12-18 v0.17 Tels 20 tests
  * require Graph::Easy 0.36
  * remove all "\n" from returned HTML output, because Mediawiki insists
    on inserting <p></p> for ever N lines or so, destroying the output
  * fix typo in graph.cfg.sample

2005-12-10 v0.16 Tels 20 tests
  * require Graph::Easy 0.35
  * require Config::Tiny 2.02
  * add PNG support via dot/neato
  * add config file support (graph.cfg)
  * put generated files into sub directories to avoid millions files-per-dir
  * quote parser error output to prevent HTML accidents

2005-12-04 v0.15 Tels 13 tests
  * require Graph::Easy 0.34
  * require Digest::SHA1 2.10
  * warn on not-installed or outdated As_svg
  * fix SVG output: create an SVG file in images/graph and embed it
    via an OBJECT tag. The file name is an SHA1 of $graph->as_txt().
  * describe cpansign --verify in README better (Thanx Mutante!)
  * skip right number of tests if Graph::Easy::As_svg is not installed
    (Thanx Mutante again!)
  * fix output on Special::Version page (escape $ to \$)

2005-11-06 v0.14 Tels 11 tests
  * require Graph::Easy 0.32
  * use utf8:decode() on incoming text, to make non-ASCII input work
  * remove debug support 

2005-10-30 v0.13 Tels 9 tests
  * require Graph::Easy 0.31
  * allow "boxart" as output mode
  * put STDOUT into ':utf8' mode (for utf-8 output)
  * do not add spurious linebreak in Special::Version page

2005-09-23 v0.12 Tels 9 tests
  * require Graph::Easy v0.30
  * fix SVG output (leading \n and empty lines were garbled by MediaWiki)
  * avoid double line numbers double line numbers in error messages
  * support the output-attribute on graphs like in "graph { output: html; }"
  * include TODO, note that SVG output still garbles the page (but less so)

2005-09-18 v0.11 Tels 7 tests
  * require Graph::Easy v0.28
  * remove unnec. "\n" after CSS and HTML code in output, this created
    ugly empty spaces below and over the graph
  * propagate error messages back so that the editing user can see them

2005-09-04 v0.10 Tels 7 tests
  * fix tests to run on Windows
  * require Graph::Easy v0.27
  * include version on the special page

2005-07-14 v0.09 Tels 7 tests
  * include our Credits on Special::Version
  * require Graph::Easy v0.22

2005-06-25 v0.08 Tels 7 tests
  * set an ABSTRACT for CPAN
  * require Graph::Easy v0.20
  * skip SVG tests if Graph::Easy::As_svg is not installed

2005-06-15 v0.07 Tels 7 tests
  * use Graph::Easy v0.19 now
  * support output format SVG
  * die() on unknown output formats

2005-03-11 v0.06 Tels 5 tests
  * add (c) note to Graph.php
  * include CHANGES
  * require Graph::Simple v0.14

2005-01-25 v0.05 Tels 5 tests
  * fix the dumb typo in Graph.php
  * graphcnv: no need to remove spaces from lines - a tad faster now

2005-01-24 v0.04 Tels 5 tests
  * rewrote as extension
  * graphcnv: takes no longer unused directory parameters

2005-01-12 v0.03 Tels 5 tests
  * further cleaning of Graph.php - and pass outputmode to graphcnv
  * graphcnv: take outputmode and return HTML or ASCII (wrapped in pre)
  * graphcnv: use Taint mode for security

2005-01-11 v0.02 Tels
  * fixed patch (relative to ., not wiki, thanx Omega!)
  * fixed README up a lot (Thanx again, Omega!)

2005-01-07 v0.01 Tels
  * first try



extensions/Graph.php
<?php
# Graph WikiMedia extension

# (c) by Tels http://bloodgate.com 2004-2006

# Takes text between <graph> </graph> tags, and runs it through the
# external script "graphcnv", which generates an ASCII, HTML or SVG
# graph from it.

$wgExtensionFunctions[] = "wfGraphExtension";
 
function wfGraphExtension() {
    global $wgParser;

    # register the extension with the WikiText parser
    # the second parameter is the callback function for processing the text between the tags

    $wgParser->setHook( "graph", "renderGraph" );
}

# for Special::Version:

$wgExtensionCredits['parserhook'][] = array(
	'name' => 'graph extension',
	'author' => 'Tels',
	'url' => 'http://wwww.bloodgate.com/perl/graph/',
	'version' => 'v0.18 using Graph::Easy v' . `perl -MGraph::Easy -e 'print \$Graph::Easy::VERSION'`,
);
 
# The callback function for converting the input text to HTML output
function renderGraph( $input ) {
    global $wgInputEncoding;

    if( !is_executable( "graph/graphcnv" ) ) {
	return "<strong class='error'><code>graph/graphcnv</code> is not executable</strong>";
    }

    $cmd = "graph/graphcnv ".  escapeshellarg($input);
    # ." ". escapeshellarg($wgInputEncoding);
    $output = `$cmd`;

    if (strlen($output) == 0) {
	return "<strong class='error'>Couldn't execute <code>graph/graphcnv</code></strong>"
               . ' See the <a alt="Graph::Easy online manual" title="Graph::Easy online manual" href="'
	       . 'http://bloodgate.com/perl/graph/manual/errors.html#1">manual</a> for help.';
    }

    return $output;
}
?>


Makefile.PL
use 5.006001;
use ExtUtils::MakeMaker;
use strict;

sub MY::metafile {
  package MY;
  my $self = shift;
  if (!eval { require YAML; 1 })
  {
    warn ("YAML not installed, 'make dist' will not override 'META.yml'.");
    return $self->SUPER::metafile_target(@_);
  };

  require YAML::Node;
  my $node = YAML::Node->new({});

  $node->{name}                 = $self->{DISTNAME};
  $node->{abstract}		= $self->{ABSTRACT};
  $node->{version}              = $self->{VERSION};
  $node->{version_from}         = $self->{VERSION_FROM};
  $node->{license}              = 'gpl';
  $node->{distribution_type}    = 'module';
  $node->{generated_by}         = "$self->{DISTNAME} version $self->{VERSION}";
  $node->{installdirs}          = 'site';
  $node->{requires}             = $self->{PREREQ_PM};

  my $dump = YAML::Dump( $node );

  $dump =~ s/^(.*)$/\t\$(NOECHO) \$(ECHO) "$1" >>META.yml/gm;
  $dump =~ s/>>META\.yml/>META.yml/;

  return "metafile:\n$dump";
}

# svg is done via dot, too.
eval { require Graph::Easy::As_svg; };

my $ver = $Graph::Easy::As_svg::VERSION;

if (!defined $ver)
  {
  print "\n************\nWARNING: ";
  print "You do not seem to have Graph::Easy::As_svg installed.\n";
  print "         SVG output (without dot) is not possible until you install it afterwards.\n\n";
  }
elsif ($ver <= 0.16)
  {
  print "\n************\n";
  print "Warning: Your Graph::Easy::As_svg version is outdated.\n";
  print "         Please upgrade it after installing this package.\n\n";
  }

WriteMakefile(
    NAME		=> 'wikimedia-graph',
    VERSION_FROM	=> 'graphcnv',
    PREREQ_PM		=> {
	'Graph::Easy' => 0.40,
	'Config::Tiny' => 2.02,
	'Digest::SHA1' => 2.10,
	},
    ABSTRACT		=> 'Integrate Graph::Easy into Wikimedia',
    AUTHOR		=> 'Tels <nospam-abuse@bloodgate.com>',
    PREREQ_FATAL	=> 1,  # can't work w/o proper prereqs
);
